File: Problems/No.345_Reverse_Vowels_of_a_String/reverse_vowels_of_a_string.cpp
class Solution {
public:

    // check if it is a vowel
    bool isVowel(char &c) {

        if ((c == 'a') || (c == 'e') || (c == 'i') || (c == 'o') || (c == 'u') ||
            (c == 'A') || (c == 'E') || (c == 'I') || (c == 'O') || (c == 'U')) return true;
        else return false;

    }

    string reverseVowels(string s) {
        
        // parameters
        int pLeft = 0, pRight = s.length()-1;

        // start computing
        while (true) {

            while ((pLeft < pRight) && !isVowel(s[pLeft])) pLeft++;
            while ((pLeft < pRight) && !isVowel(s[pRight])) pRight--;

            if (pLeft < pRight) std::swap(s[pLeft++], s[pRight--]);
            else break;

        }

        return s;

    }
};

File: Problems/No.345_Reverse_Vowels_of_a_String/reverse_vowels_of_a_string.py
class Solution(object):
    def reverseVowels(self, s):
        """
        :type s: str
        :rtype: str
        """

        # parameters
        pLeft, pRight = 0, len(s)-1
        vowels = ['a', 'e', 'i', 'o', 'u', 'A', 'E', 'I', 'O', 'U']
        s_list = list(s)

        # start computing
        while (True):
            while ((pLeft < pRight) and (s_list[pLeft] not in vowels)):
                pLeft += 1
            while ((pLeft < pRight) and (s_list[pRight] not in vowels)):
                pRight -= 1
            if (pLeft < pRight):
                s_list[pLeft], s_list[pRight] = s_list[pRight], s_list[pLeft]
                pLeft += 1
                pRight -= 1
            else:
                break
        
        return ''.join(s_list)
